---
title: "Genome-Wide Search Algorithms for Identifying Dynamic Gene Co-Expression via Bayesian Variable Selection"
author: "Zichen Ma"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(MCMCpack)
#library(dplyr)

set.seed(10)
```

### 1. Introduction

This document provides a tutorial for the R functions `spsl()` and `cspsl()` that fit the spike-and-slab model (SPSL) and the correlated spike-and-slab model (C-SPSL), respectively, in analyzing gene co-expression patterns. The methods are based on Zhang et al. (under revision, *Statistics in Medicine*). Given a set of gene expression data, traditional methods for gene co-expression analysis such as the conditional normal model proposed in Ho et al. (2011) require exhaustive search across all pairs of genes, which may be expensive when the set of genes is large. SPSL and C-SPSL utilize the spike-and-slab prior from Bayesian variable selection. 

### 2. Example

The `spsl()` function and the `cspsl()` function are comparable in terms of the syntax and the output. These functions, together with the function `summary.stats()` which calculates the Bayes factor, are included in the source file `Functions.R` and can be called by using the `source()` function.

```{r}
source("Functions.R")
```

#### 2.1. Fit the SPSL model 

##### 2.1.1. Posterior of $\tau_1$ via `spsl()`

We now describe the `spsl()` function in detail. The `spsl()` function fits a Bayesian spike-and-slab model according to Zhang et al. (2023) that analyzes the gene-gene interaction. The main input parameters and default values (if applicable) of the function include

* `data`: an $n\times q$ matrix or data frame of gene expression levels, where $n$ is the sample size and $q$ is the number of genes.
* `z`: a vector of length $n$ of the modulating factor.
* `iterations = 10000`: the number of MCMC iterations.
* `burnIn = 5000`: length of initial burn-in.
* `chains = 3`: number of Markov chains, mainly used for convergence diagnostics.

To give an example, consider the following data set, which is taken from the BRCA-US project at The Cancer Genome Atlas (TCGA). The data frame `demo` includes RNA-sequencing data of $q=4$ different genes across $n=457$ subjects. The vector `z` is the modulating factor - the survival time of breast cancer standardized to the interval $(0,1)$. A brief summary of the data frame `demo` is also shown below.

```{r}
demo <- read.table("Xmat_demo.txt",header=T)
z <- read.table("Zmat.txt",header=F)$V1

summary(demo)

summary(z)
```

We now call the `spsl()` function with the following input, and save the output as an object `out.spsl`.

```{r}
out.spsl = spsl(data=demo, z, iterations = 5000, burnIn = 1000, chains = 1)
class(out.spsl)
```

The output `out.spsl` is a list of length equal to the number of chains. In this case, since we set `chains = 1`, the output is a list of length 1. The element can be accessed using `out.spsl[[1]]`. The element itself is a list containing the MCMC draws of model parameters in the SPSL model: $\tau_0$, $\tau_1$, $\phi$, and $w$. 

```{r}
class(out.spsl[[1]])

names(out.spsl[[1]])
```

Specific parameters can be accessed using `$`. Of particular interest is the posterior of $\tau_1$, which is a matrix of $\binom{4}{2}=6$ columns. The number of rows is 4000, the length of the Markov chain after burn-in.

```{r}
out.spsl.tau1 = out.spsl[[1]]$tau1
dim(out.spsl.tau1)
```

Further, we can obtain point and interval estimates based on `out.spsl.tau1`. The following output provides the posterior median together with a 95\% credible interval of $\tau_{11},\tau_{12},\ldots,\tau_{16}$.

```{r}
round(apply(out.spsl.tau1,2,quantile,probs=c(.025, 0.5, 0.975)),3)
```

##### 2.1.2. The Bayes factor via `summary.stats()`

Moreover, for $j=1,2,\ldots,6$, the function `summary.stats()` provides the local Bayes factor for testing 
$$
H_{0j}:\vert\tau_{1j}\vert \leq \epsilon \text{  versus  }H_{1j}:\vert\tau_{1j}\vert > \epsilon.
$$

The input parameters and default values (if applicable) of the function include

* `tau1.post`: a matrix of MCMC draws of the posterior of **all** $\tau_1$ parameters
* `e=0.4`: the value of $\epsilon$. Default set at 0.4 to enforce $1:1$ prior odds of $H_{0j}$ against $H_{1j}$.
* `alpha=0.05`: the upper bound of the expected false discovery rate (FDR).

The output is a list of two elements:

* `bf`: a vector of the local Bayes factors
* `r.opt`: a scalar of the optimal cutoff for the Bayes factor which ensures that the upper bound set at `alpha` is met.

```{r}
summary.stats(tau1.post=out.spsl.tau1, e=0.4, alpha=0.05)
```


#### 2.2. Fit the correlated spike-and-slab model using `cspsl()`

The `cspsl()` function is mostly comparable to the `spsl()` function in that the two functions have the same syntax and input parameters. In addition to the output of `spsl()`, the `cspsl()` function contains the posterior draws of the parameter $c$ in the C-SPSL model, which does not exist in SPSL.

Shown below is the output of the $\tau_1$ parameters, including the posterior median, 95\% credible interval, and the local Bayes factors. Note that the results are fairly similar to the results based on the SPSL model.

```{r}
out.cspsl = spsl(data=demo, z, iterations = 5000, burnIn = 1000, chains = 1)
out.cspsl.tau1 = out.cspsl[[1]]$tau1

round(apply(out.cspsl.tau1,2,quantile,probs=c(.025, 0.5, 0.975)),3)

summary.stats(tau1.post=out.cspsl.tau1, e=0.4, alpha=0.05)
```